### PR TITLE
PARQUET-2067: [C++][Parquet] Fix Parquet null count stats for enclosing null lists

### PR DESCRIPTION
**Pull request by tachyonwill**
For repeated columns, the null count should include cases where the
enclosing list(s) is null of empty. For example `[[0, 1], None, [2,
None, 3]]` should have a null count of 2.

Co-authored-by: Micah Kornfield <micahk@google.com>

### CODE REVIEW

File: cpp/src/parquet/arrow/arrow_reader_writer_test.cc [hunk 1]
```diff
@@ -3896,6 +3896,41 @@ INSTANTIATE_TEST_SUITE_P(WriteDictionary, TestArrowWriteDictionary,
                          ::testing::Values(ParquetDataPageVersion::V1,
                                            ParquetDataPageVersion::V2));
 
+TEST(TestArrowWriteNestedDictionary, TestNullCount) {
+  std::shared_ptr<::arrow::Array> dictionary = ArrayFromJSON(::arrow::utf8(), R"(["z"])");
```
Comment (emkornfield) [inline cpp/src/parquet/arrow/arrow_reader_writer_test.cc:3900]: any reason to use dictionary here?

Comment (tachyonwill) [reply cpp/src/parquet/arrow/arrow_reader_writer_test.cc:3900]: If you are asking why I added a dictionary specific test, it is to test the changes to WriteArrowDictionary; the other tests were not sufficient.

File: cpp/src/parquet/statistics.h [hunk 2]
```diff
@@ -282,8 +282,8 @@ class TypedStatistics : public Statistics {
 
   /// \brief Batch statistics update with supplied validity bitmap
   virtual void UpdateSpaced(const T* values, const uint8_t* valid_bits,
```
Comment (pitrou) [inline cpp/src/parquet/statistics.h:284]: Can you add a description of the parameters to the docstring? It is not immediately obvious what `num_spaced_values` is...

File: cpp/src/parquet/arrow/arrow_reader_writer_test.cc [hunk 3]
```diff
@@ -3896,6 +3896,41 @@ INSTANTIATE_TEST_SUITE_P(WriteDictionary, TestArrowWriteDictionary,
...
+  std::shared_ptr<::arrow::Array> indices =
+      ArrayFromJSON(::arrow::int32(), R"([0, null, 0])");
+  std::shared_ptr<::arrow::Array> offsets =
+      ArrayFromJSON(::arrow::int32(), R"([null, 0, null, null, null, 3])");
+  ASSERT_OK_AND_ASSIGN(std::shared_ptr<::arrow::Array> dict_encoded,
+                       ::arrow::DictionaryArray::FromArrays(indices, dictionary));
+  // values looks like [null, ["z", null, "z"], null, null, null]
+  ASSERT_OK_AND_ASSIGN(
+      std::shared_ptr<Array> values,
+      ::arrow::ListArray::FromArrays(*offsets, *dict_encoded, default_memory_pool()));
```
Comment (pitrou) [inline cpp/src/parquet/arrow/arrow_reader_writer_test.cc:3910]: You can directly create of list-of-dictionary array using `ArrayFromJSON`. For example:
```c++
auto type = list(dictionary(int32(), utf8()));
auto values = ArrayFromJSON(type, R"([null, ["z", null, "z"], null, null, null])");
```

File: cpp/src/parquet/arrow/arrow_reader_writer_test.cc [hunk 4]
```diff
@@ -3896,6 +3896,41 @@ INSTANTIATE_TEST_SUITE_P(WriteDictionary, TestArrowWriteDictionary,
...
+
+  std::shared_ptr<::arrow::ResizableBuffer> serialized_data = AllocateBuffer();
+  auto out_stream = std::make_shared<::arrow::io::BufferOutputStream>(serialized_data);
+  std::unique_ptr<FileWriter> writer;
+  ASSERT_OK(FileWriter::Open(*table->schema(), default_memory_pool(), out_stream,
+                             default_writer_properties(),
+                             default_arrow_writer_properties(), &writer));
+  ASSERT_OK(writer->WriteTable(*table, std::numeric_limits<int64_t>::max()));
+  ASSERT_OK(writer->Close());
+  if (!out_stream->closed()) {
```
Comment (pitrou) [inline cpp/src/parquet/arrow/arrow_reader_writer_test.cc:3923]: It is probably not necessary to test for `out_stream->closed()`, is it?

File: cpp/src/parquet/arrow/arrow_reader_writer_test.cc [hunk 5]
```diff
@@ -4191,5 +4226,107 @@ TEST(TestArrowReadDeltaEncoding, DeltaBinaryPacked) {
...
+  EXPECT_EQ(stats->EncodeMax(), GetParam().expected_max);
+}
+
+INSTANTIATE_TEST_SUITE_P(
+    StatsTests, ParameterizedStatisticsTest,
+    ::testing::Values(
+        StatisticsTestParam{/*table=*/Table::Make(
+                                ::arrow::schema({::arrow::field("a", ::arrow::utf8())}),
+                                {ArrayFromJSON(::arrow::utf8(), R"(["1", null, "3"])")}),
+                            /*expected=*/1, /* empty list counts as null as well */
```
Comment (pitrou) [inline cpp/src/parquet/arrow/arrow_reader_writer_test.cc:4280]: Should be `/*expected_null_count=*/`

File: cpp/src/parquet/arrow/arrow_reader_writer_test.cc [hunk 6]
```diff
@@ -4191,5 +4226,107 @@ TEST(TestArrowReadDeltaEncoding, DeltaBinaryPacked) {
 }
 #endif
 
+struct StatisticsTestParam {
```
Comment (pitrou) [inline cpp/src/parquet/arrow/arrow_reader_writer_test.cc:4229]: Can you perhaps create a separate `statistics_test.cc`? This test file is getting much too long already.

File: cpp/src/parquet/arrow/arrow_reader_writer_test.cc [hunk 7]
```diff
@@ -4191,5 +4226,107 @@ TEST(TestArrowReadDeltaEncoding, DeltaBinaryPacked) {
...
+  std::string many_empty_lists = "[";
+  for (int i = 0; i < 2000; ++i) {
+    many_empty_lists += "[],";
+  }
+  many_empty_lists += "[1,2,3,4,5,6,7,8,null]]";
+  return many_empty_lists;
+}
+
+TEST_P(ParameterizedStatisticsTest, NoNullCountWrittenForRepeatedFields) {
+  std::shared_ptr<::arrow::ResizableBuffer> serialized_data = AllocateBuffer();
```
Comment (pitrou) [inline cpp/src/parquet/arrow/arrow_reader_writer_test.cc:4251]: Would be nice to add a comment with a reference to the JIRA ticket.

File: cpp/src/parquet/statistics.h (hunk 2 above)
Comment (tachyonwill) [reply cpp/src/parquet/statistics.h:294]: Done.

File: cpp/src/parquet/arrow/arrow_reader_writer_test.cc (hunk 3 above)
Comment (tachyonwill) [reply cpp/src/parquet/arrow/arrow_reader_writer_test.cc]: done.

File: cpp/src/parquet/arrow/arrow_reader_writer_test.cc (hunk 5 above)
Comment (tachyonwill) [reply cpp/src/parquet/arrow/arrow_reader_writer_test.cc]: done.

File: cpp/src/parquet/arrow/arrow_reader_writer_test.cc (hunk 7 above)
Comment (tachyonwill) [reply cpp/src/parquet/arrow/arrow_reader_writer_test.cc]: done.

File: cpp/src/parquet/arrow/arrow_reader_writer_test.cc (hunk 6 above)
Comment (tachyonwill) [reply cpp/src/parquet/arrow/arrow_reader_writer_test.cc]: created arrow_statistics_test.cc

File: cpp/src/parquet/arrow/arrow_reader_writer_test.cc (hunk 4 above)
Comment (tachyonwill) [reply cpp/src/parquet/arrow/arrow_reader_writer_test.cc]: made it unconditional.

File: cpp/src/parquet/arrow/arrow_statistics_test.cc [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+using arrow::Table;
+
+using arrow::io::BufferReader;
+
+namespace parquet {
+namespace arrow {
+
+// PARQUET-2067: Tests that nulls from parent fields are included in null statistics when
+// array is dictionary encoded.
+TEST(TestArrowWriteNestedDictionary, TestNullCount) {
```
Comment (pitrou) [inline cpp/src/parquet/arrow/arrow_statistics_test.cc:44]: Can this be folded into ParameterizedStatisticsTest below? It doesn't seem there's any specific code.

File: cpp/src/parquet/statistics.h (hunk 2 above)
Comment (pitrou) [reply cpp/src/parquet/statistics.h:294]: Thank you!

**Review by pitrou — COMMENTED** [review]
Thanks! Just one more suggestion.

File: cpp/src/parquet/arrow/arrow_statistics_test.cc (hunk 8 above)
Comment (tachyonwill) [reply cpp/src/parquet/arrow/arrow_statistics_test.cc]: done

**Review by pitrou — APPROVED** [review]
Thank you @tachyonwill

**Pull request closed without merging**